src/client/serialize/hdf_serialize.c

```diff
@@ -0,0 +1,1034 @@
+#include <stdio.h>
+#include <hdf5.h>
+#include <daos.h>
+#include <daos_cont.h>
+
+/* for user attr dataset */
+struct usr_attr {
+	char	*attr_name;
+	hvl_t	attr_val;
+};
+
+static int
+cont_serialize_prop_acl(hid_t *file_id, struct daos_prop_entry *entry,
+			const char *prop_str)
+{
+
+	int		rc = 0;
+	hid_t		status = 0;
+	struct daos_acl	*acl = NULL;
+	char		**acl_strs = NULL;
+	size_t		len_acl = 0;
+	hsize_t		attr_dims[1];
+	hid_t		attr_dtype;
+	hid_t		attr_dspace;
+	hid_t		usr_attr;
+
+
+	if (entry == NULL || entry->dpe_val_ptr == NULL) {
+		goto out;
+	}
+
+	/* convert acl to list of strings */
+	acl = (struct daos_acl *)entry->dpe_val_ptr;
+	rc = daos_acl_to_strs(acl, &acl_strs, &len_acl);
+	if (rc != 0) {
+		fprintf(stderr, "failed to convert acl to strs");
+		goto out;
+	}
+	attr_dims[0] = len_acl;
+	attr_dtype = H5Tcopy(H5T_C_S1);
+	if (attr_dtype < 0) {
+		fprintf(stderr, "failed to create acl type");
+		rc = 1;
+		goto out;
+	}
+	status = H5Tset_size(attr_dtype, H5T_VARIABLE);
+	if (status < 0) {
+		fprintf(stderr, "failed to set acl dtype size");
+		rc = 1;
+		goto out;
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		fprintf(stderr, "failed to create version attribute");
+		rc = 1;
+		goto out;
+	}
+	usr_attr = H5Acreate2(*file_id, prop_str, attr_dtype, attr_dspace,
+			      H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		fprintf(stderr, "failed to create attribute");
+		rc = 1;
+		goto out;
+	}
+	status = H5Awrite(usr_attr, attr_dtype, acl_strs);
+	if (status < 0) {
+		fprintf(stderr, "failed to write attribute");
+		rc = 1;
+		goto out;
+	}
+	status = H5Aclose(usr_attr);
+	if (status < 0) {
+		fprintf(stderr, "failed to close attribute");
+		rc = 1;
+		goto out;
+	}
+	status = H5Tclose(attr_dtype);
+	if (status < 0) {
+		fprintf(stderr, "failed to close dtype");
+		rc = 1;
+		goto out;
+	}
+	status = H5Sclose(attr_dspace);
+	if (status < 0) {
+		fprintf(stderr, "failed to close dspace");
+		rc = 1;
+		goto out;
+	}
+out:
```

Comment:
you leak a lot of HDF5 objects on failure. should make sure those are closed depending where you fail.
this applies to most functions in this file.